### PR TITLE
Eager PyroSample evaluation when PyroModule context is active

### PR DESCRIPTION
**Pull request by eb8680**

### CODE REVIEW

File: pyro/nn/module.py [hunk 1]
```diff
@@ -180,10 +180,20 @@ def __init__(self):
...
         self.used = True
 
     def __exit__(self, type, value, traceback):
+        if not (self.active - 1):
+            # attr statements below depend on bool(self.active) is True
+            for name, mod in self._local_samples.items():
+                const_value = getattr(mod, name)
+                assert not isinstance(const_value, torch.nn.Parameter)
+                delattr(mod, name)
+                setattr(mod, name, const_value)
```
Comment (eb8680) [inline pyro/nn/module.py:195]: Use `register_buffer` here (or in `__setattr__`), and detach `const_value`

**Comment by eb8680** [comment]
Closing in favor of a different approach that doesn't change `PyroSample` behavior following discussion with @fritzo
